088-rebalance-progress-status.md

```diff
@@ -0,0 +1,392 @@
+
+#### State: `ProposalReady`
+
+Since the rebalance has not started at this point, we do not have the data needed to easily provide an accurate value for this field.
+Therefore, we omit this field in the `ConfigMap` of the `KafkaRebalance` resource for this state.
+
+#### State: `Rebalancing`
+
+The estimated time it will take for the ongoing rebalance to complete based on the average rate of data transfer.
+
+$$
+r = \frac{D_{moved}}{T_{\text{current}} - T_{\text{start}}}
+$$
+
+$$
+ETC = \frac{D_{total} - D_{moved}}{r \cdot 60000}
+$$
+
+**Notes:**
+- $r$: The average rate of data transfer in megabytes per second, calculated as the ratio of finished data movement to the elapsed time.
+- $D_{moved}$: The number of megabytes already moved by the rebalance, provided by `finishedDataMovement` field from the [/kafkacruisecontrol/state?substates=executor](#field-executorstate.json) REST API endpoint.
+- $T_{\text{current}}$: The current time when the estimate is being calculated, expressed as the number of milliseconds since the Unix epoch.
+- $T_{\text{start}}$: The time when the rebalance task was triggered, expressed as the number of milliseconds since the Unix epoch, extracted from the `triggeredTaskReason` field from the [/kafkacruisecontrol/state?substates=executor](#field-executorstate.json) REST API endpoint.
+- $D_{total}$: The total number of megabytes planned to be moved for the rebalance, provided by `totalDataToMove` field from the [/kafkacruisecontrol/state?substates=executor](#field-executorstate.json) REST API endpoint.
+- $ETC$: The estimated time to completion in minutes based on the average rate of data transfer, the value of the `estimatedTimeToCompletionInMinutes` field.
+
+#### State: `Stopped`
+
+Once a rebalance has been stopped, it cannot be resumed. 
+Therefore, there is no `estimatedTimeToCompletionInMinutes` for a stopped rebalance so we omit this field in the `ConfigMap` of the `KafkaRebalance` resource for this state.
+To move from the `Stopped` state, a user must refresh the `KafkaRebalance` resource, the `rebalanceProgressConfigMap` will then be updated by the operator upon the next state change.
+
+#### State: `NotReady`
+
+Once a rebalance has been interrupted or completed with errors, it cannot be resumed.
+Therefore, there is no `estimatedTimeToCompletionInMinutes` for the rebalance so we omit this field in the `ConfigMap` of the `KafkaRebalance` resource for this state.
+To move from the `NotReady` state, a user must refresh the `KafkaRebalance` resource, the `rebalanceProgressConfigMap` will then be updated by the operator upon the next state change.
+
+#### State: `Ready`
+
+The rebalance is complete so we hardcode the value to `0`
+This emphasizes that the rebalance is complete and helps clear up ambiguity surrounding what the `Ready` state means in the `KafkaRebalance` resource.
+
+### Field: `completedByteMovementPercentage`
```

Comment:
The "byte" part of this smells wrong, why not "data"?

Reply:
It was raised in an earlier comment that "data" movement could be misinterpreted as partition movement instead of the byte movement. Naming this field with "byte" removes any ambiguity as to what is being measured

Reply:
Not sure it would be easy to mix up data and partitions personally, but ok. I still think it would be better to have `completedDataMovementPercentage` and `completedPartitionMovementPercentage` then you can't mix them up.

Reply:
>  I still think it would be better to have `completedDataMovementPercentage` and `completedPartitionMovementPercentage` then you can't mix them up.

I was worried that information from these two fields would be too similar, therefore, I was hoping to only supply one of them. However, I want the field name(s) to be as clear as possible to everyone. I am open to including both, it was something which @katheris raised in an earlier review too.

I would be interested in what @scholzj thinks of this.

Reply:
Discussed this with Jakub on a call today, even if we were to include `completedPartitionMovementPercentage` the field in the proposal, the "Data" in `completedDataMovementPercentage` field still seems that it could be potentially ambiguous. 

If using the field name `completedByteMovementPercentage` is not a blocker for anyone, let's stick with this for now. We can always add the `completedPartitionMovementPercentage` field in a later proposal too!

CC @katheris would still like your opinion on this too ^^^